### PR TITLE
[feat] add M4C model for TextVQA

### PR DESCRIPTION
**Pull request by ronghanghu**
Merge the M4C model (https://arxiv.org/pdf/1911.06258.pdf) for TextVQA into Pythia.

Summary of changes:
* Adding `README.md` under `projects/M4C`
* Adding new models: M4C under `pythia/models/m4c.py`
* Adding new dataset classes: `m4c_textvqa`, `m4c_stvqa`, and `m4c_ocrvqa` under `pythia/datasets/vqa/`
* Adding new config files under `configs/vqa`
* Adding new processors, metrics and losses for M4C training and evaluation.
* Adding other utilities (such as PHOC feature extraction).

Introducing new dependencies (added to `requirements.txt`):
* `pytorch-transformers`  
* `editdistance`

## M4C for the TextVQA Task
* R. Hu, A. Singh, T. Darrell, M. Rohrbach, *Iterative Answer Prediction with Pointer-Augmented Multimodal Transformers for TextVQA*. arXiv preprint arXiv:1911.06258, 2019 ([PDF](https://arxiv.org/pdf/1911.06258.pdf))
```
@article{hu2019iterative,
  title={Iterative Answer Prediction with Pointer-Augmented Multimodal Transformers for TextVQA},
  author={Hu, Ronghang and Singh, Amanpreet and Darrell, Trevor and Rohrbach, Marcus},
  journal={arXiv preprint arXiv:1911.06258},
  year={2019}
}
```

## Vocabs, ImDBs and Features:

| Datasets      | M4C Vocabs | M4C ImDBs | Object Faster R-CNN Features | OCR Faster R-CNN Features |
|--------------|-----|-----|-----------------------------------------------------------------------------------|---------------------------------------------------------------------------------|
| TextVQA      | [All Vocabs](https://dl.fbaipublicfiles.com/pythia/m4c/data/m4c_vocabs.tar.gz) | [TextVQA ImDB](https://dl.fbaipublicfiles.com/pythia/m4c/data/imdb/m4c_textvqa.tar.gz) | [OpenImages](https://dl.fbaipublicfiles.com/pythia/features/open_images.tar.gz) | [TextVQA Rosetta-en OCRs](https://dl.fbaipublicfiles.com/pythia/m4c/data/m4c_textvqa_ocr_en_frcn_features.tar.gz), [TextVQA Rosetta-ml OCRs](https://dl.fbaipublicfiles.com/pythia/m4c/data/m4c_textvqa_ocr_ml_frcn_features.tar.gz) |
| ST-VQA      | [All Vocabs](https://dl.fbaipublicfiles.com/pythia/m4c/data/m4c_vocabs.tar.gz) | [ST-VQA ImDB](https://dl.fbaipublicfiles.com/pythia/m4c/data/imdb/m4c_stvqa.tar.gz) | [ST-VQA Objects](https://dl.fbaipublicfiles.com/pythia/m4c/data/m4c_stvqa_obj_frcn_features.tar.gz) | [ST-VQA Rosetta-en OCRs](https://dl.fbaipublicfiles.com/pythia/m4c/data/m4c_stvqa_ocr_en_frcn_features.tar.gz) |
| OCR-VQA      | [All Vocabs](https://dl.fbaipublicfiles.com/pythia/m4c/data/m4c_vocabs.tar.gz) | [OCR-VQA ImDB](https://dl.fbaipublicfiles.com/pythia/m4c/data/imdb/m4c_ocrvqa.tar.gz) | [OCR-VQA Objects](https://dl.fbaipublicfiles.com/pythia/m4c/data/m4c_ocrvqa_obj_frcn_features.tar.gz) | [OCR-VQA Rosetta-en OCRs](https://dl.fbaipublicfiles.com/pythia/m4c/data/m4c_ocrvqa_ocr_en_frcn_features.tar.gz) |

## Pretrained models:

| Datasets  | Configs (under `configs/vqa/`)         | Pretrained Models | Metrics                     | Notes                         |
|--------|------------------|----------------------------|-------------------------------|-------------------------------|
| TextVQA (`m4c_textvqa`) | `m4c_textvqa/m4c_with_stvqa.yml` | [`download`](https://dl.fbaipublicfiles.com/pythia/m4c/m4c_release_models/m4c_textvqa/m4c_textvqa_m4c_with_stvqa.ckpt) | val accuracy - 40.55%; test accuracy - 40.46% | Rosetta-en OCRs; ST-VQA as additional data |
| TextVQA (`m4c_textvqa`) | `m4c_textvqa/m4c.yml` | [`download`](https://dl.fbaipublicfiles.com/pythia/m4c/m4c_release_models/m4c_textvqa/m4c_textvqa_m4c.ckpt) | val accuracy - 39.40%; test accuracy - 39.01% | Rosetta-en OCRs |
| TextVQA (`m4c_textvqa`) | `m4c_textvqa/m4c_ocr_ml.yml` | [`download`](https://dl.fbaipublicfiles.com/pythia/m4c/m4c_release_models/m4c_textvqa/m4c_textvqa_m4c_ocr_ml.ckpt) | val accuracy - 37.06% | Rosetta-ml OCRs |
| ST-VQA (`m4c_stvqa`)  | `m4c_stvqa/m4c.yml` | [`download`](https://dl.fbaipublicfiles.com/pythia/m4c/m4c_release_models/m4c_stvqa/m4c_stvqa_m4c.ckpt) | val ANLS - 0.472 (accuracy - 38.05%); test ANLS - 0.462 | Rosetta-en OCRs |
| OCR-VQA (`m4c_ocrvqa`) | `m4c_ocrvqa/m4c.yml` | [`download`](https://dl.fbaipublicfiles.com/pythia/m4c/m4c_release_models/m4c_ocrvqa/m4c_ocrvqa_m4c.ckpt) | val accuracy - 63.52%; test accuracy - 63.87% | Rosetta-en OCRs |

### CODE REVIEW

**Comment by apsdehal** [comment]
Any plans to fix the build?

**Comment by ronghanghu** [comment]
> Any plans to fix the build?

Yes! After finishing the captioning experiments, I'm addressing the issues you mentioned offline last week (including the CI errors above), and will update this PR.

File: pythia/utils/m4c_evaluators.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+        out_text = " ".join(out_text)
+        return out_text
+
+    def __call__(self, item):
+        item = self.word_tokenize(item)
+        item = item.replace("\n", " ").replace("\t", " ").strip()
+        item = self.process_punctuation(item)
+        item = self.process_digit_article(item)
+        return item
+
```
Comment (vedanuj) [inline pythia/utils/m4c_evaluators.py]: Please reuse the existing code and remove this duplication.

Comment (ronghanghu) [reply pythia/utils/m4c_evaluators.py]: @vedanuj Thanks for the comments! I'll address these issues later this week.

**Comment by ronghanghu** [comment]
@vedanuj thanks for your review!

> Any plans to fix the build?

It's fixed now, by `pytest==5.2.0` in requirements.txt

> Please make sure all new files have the required copyright headers.

The header `Copyright (c) Facebook, Inc. and its affiliates.` is added to the new files.

> Please reuse the existing code and remove this duplication.

I found that this is non-trivial to fix. This PR is made against v0.4. However, the `EvalAIAnswerProcessor` only exists in the current master branch. I tried rebasing against master. However, I found that v0.4 has a major commit ahead of master (https://github.com/facebookresearch/pythia/commit/926d3b0d2340c1d65730764ee8f84c50837cdda8) that address multitasking https://github.com/facebookresearch/pythia/pull/173, which cases a lot of rebase conflict. (Note that this PR for the M4C model is built to be compatible with https://github.com/facebookresearch/pythia/pull/173, which I believe eventually will appear in master)

Do you have suggestions on how to proceed here?

File: requirements.txt [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
+tqdm>=4.19
 demjson>=2.2
 torchtext>=0.2
 GitPython>=2.1
 PyYAML>=3.11
-pytest==3.3.2
+pytest==5.2.0
 requests==2.21.0
-fasttext==0.9.1
+fastText
```
Comment (apsdehal) [inline requirements.txt]: Any particular reason that we are not using a specific version here?

File: requirements.txt [hunk 3]
```diff
@@ -1,13 +1,15 @@
...
 nltk==3.4.1
+pytorch-transformers==1.2.0
+editdistance
```
Comment (apsdehal) [inline requirements.txt]: Same here?

File: pythia/utils/phoc/src/cphoc.c [hunk 4]
```diff
@@ -0,0 +1,146 @@
+// C implementation of the PHOC respresentation. Converts a string into a PHOC feature vector
```
Comment (apsdehal) [inline pythia/utils/phoc/src/cphoc.c]: We should move M4C specific utils to a folder `utils/m4c_utils`.

File: pythia/utils/objects_to_byte_tensor.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
+
```
Comment (apsdehal) [inline pythia/utils/objects_to_byte_tensor.py]: This should go inside distributed utils.

File: pythia/common/defaults/configs/datasets/vqa/m4c_stvqa.yml [hunk 6]
```diff
@@ -0,0 +1,22 @@
+dataset_attributes:
```
Comment (apsdehal) [inline pythia/common/defaults/configs/datasets/vqa/m4c_stvqa.yml]: Configs currently have a lot of replication and not fully utilizing the power of inheritance in our configuration system.

File: projects/M4C/scripts/extract_ocr_frcn_feature.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+import numpy as np
+import tqdm
+import argparse
+import torch
+from PIL import Image
+import cv2
+
+# install `vqa-maskrcnn-benchmark` from
+# https://github.com/ronghanghu/vqa-maskrcnn-benchmark-m4c
+import sys; sys.path.append('/private/home/ronghanghu/workspace/vqa-maskrcnn-benchmark')  # NoQA
```
Comment (vedanuj) [inline projects/M4C/scripts/extract_ocr_frcn_feature.py]: Is there some specific change done for m4c? If not can the `vqa-maskrcnn-benchmark` repo be used here?  Because it is already being used in feature extraction scripts pythia provides here `pythia/scripts/features/extract_features_vmb.py`

File: .gitignore [hunk 8]
```diff
@@ -24,3 +24,5 @@ eggs/
 *.egg
 .DS_Store
 .vscode/*
+*.so
+*-checkpoint.ipynb
```
Comment (vedanuj) [inline .gitignore]: Are `.ipynb` files generated by any of the scripts added here? If not please remove it.

File: configs/captioning/m4c_textcaps/m4c_captioner_coco_textcaps_joint.yml [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+      - open_images/detectron_fix_100/fc6/train,m4c_textvqa_ocr_en_frcn_features/train_images
+      - open_images/detectron_fix_100/fc6/train,m4c_textvqa_ocr_en_frcn_features/train_images
+      - open_images/detectron_fix_100/fc6/train,m4c_textvqa_ocr_en_frcn_features/train_images
+      - open_images/detectron_fix_100/fc6/train,m4c_textvqa_ocr_en_frcn_features/train_images
+      - open_images/detectron_fix_100/fc6/train,m4c_textvqa_ocr_en_frcn_features/train_images
+      - open_images/detectron_fix_100/fc6/train,m4c_textvqa_ocr_en_frcn_features/train_images
+      - open_images/detectron_fix_100/fc6/train,m4c_textvqa_ocr_en_frcn_features/train_images
+      - open_images/detectron_fix_100/fc6/train,m4c_textvqa_ocr_en_frcn_features/train_images
+      - open_images/detectron_fix_100/fc6/train,m4c_textvqa_ocr_en_frcn_features/train_images
+      - open_images/detectron_fix_100/fc6/train,m4c_textvqa_ocr_en_frcn_features/train_images
```
Comment (vedanuj) [inline configs/captioning/m4c_textcaps/m4c_captioner_coco_textcaps_joint.yml]: This behaviour should be configurable from code in future. This looks very messy now.

File: projects/M4C/scripts/extract_ocr_frcn_feature.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+    im = cv2.resize(
+        im,
+        None,
+        None,
+        fx=im_scale,
+        fy=im_scale,
+        interpolation=cv2.INTER_LINEAR
+    )
+    img = torch.from_numpy(im).permute(2, 0, 1)
+    return img, im_scale
```
Comment (vedanuj) [inline projects/M4C/scripts/extract_ocr_frcn_feature.py]: Please check if code can be reused between these and  `pythia/scripts/features/extract_features_vmb.py`

File: pythia/datasets/processors.py [hunk 11]
```diff
@@ -895,3 +899,288 @@ def __call__(self, item):
...
+    def __call__(self, item):
+        blob = item["blob"]
+        final_blob = np.zeros((self.max_length,) + blob.shape[1:], blob.dtype)
+        final_blob[:len(blob)] = blob[:len(final_blob)]
+
+        return {"blob": torch.from_numpy(final_blob)}
+
+
+@registry.register_processor("bert_tokenizer")
+class BertTokenizerProcessor(BaseProcessor):
```
Comment (vedanuj) [inline pythia/datasets/processors.py]: @apsdehal We should check if this is same or different from the BertTokenizer we have internally when merging.

File: pythia/modules/losses.py [hunk 12]
```diff
@@ -160,7 +160,7 @@ def forward(self, sample_list, model_output, *args, **kwargs):
         if loss.dim() == 0:
             loss = loss.view(1)
 
-            key = "{}/{}/{}".format(
```
Comment (vedanuj) [inline pythia/modules/losses.py]: Was this a bug earlier?

File: pythia/utils/checkpoint.py [hunk 13]
```diff
@@ -42,7 +42,7 @@ def __init__(self, trainer):
 
         self.models_foldername = os.path.join(self.ckpt_foldername, "models")
         if not os.path.exists(self.models_foldername):
-            os.makedirs(self.models_foldername)
+            os.makedirs(self.models_foldername, exist_ok=True)
```
Comment (vedanuj) [inline pythia/utils/checkpoint.py]: Can we not change this? This might have dangerous consequences where we overwrite trained models by mistake.

File: requirements.txt [hunk 14]
```diff
@@ -1,13 +1,15 @@
...
+tensorboardX>=1.2
 numpy>=1.14
-tqdm==4.19.9
+tqdm>=4.19
 demjson>=2.2
 torchtext>=0.2
 GitPython>=2.1
 PyYAML>=3.11
-pytest==3.3.2
+pytest==5.2.0
```
Comment (vedanuj) [inline requirements.txt]: Are these changes necessary for this PR? If not please remove it from this.

File: pythia/utils/checkpoint.py (hunk 13 above)
Comment (apsdehal) [reply pythia/utils/checkpoint.py]: This is fine. This is an issue in case of distributed training which is already fixed in our dev branch. If you see the `if` statement above you will understand. This happens when there is a race condition and one of the jobs have already made the folder, then the whole thing fails. That's why `exist_ok=True` needs. `exist_ok` doesn't overwrite, it just doesn't throw an error if the folder is already there.

File: pythia/modules/losses.py (hunk 12 above)
Comment (apsdehal) [reply pythia/modules/losses.py]: Yeah, I noticed this and fixed it in the rebase. It was already fixed in dev branch,

File: pythia/datasets/processors.py (hunk 11 above)
Comment (apsdehal) [reply pythia/datasets/processors.py]: Ours is inside `processors/bert`.

File: projects/M4C/scripts/extract_ocr_frcn_feature.py (hunk 7 above)
Comment (apsdehal) [reply projects/M4C/scripts/extract_ocr_frcn_feature.py]: Yeah, this needs a separate look.

File: requirements.txt (hunk 14 above)
Comment (apsdehal) [reply requirements.txt]: I removed the changes with `>=` as they can be breaking in nature. Rest I accepted and then updated with our changes as of now.

**Comment by apsdehal** [comment]
@ronghanghu This has landed internally on dev. Please make further PRs on that branch. This should be automatically closed once it lands in master.

File: pythia/utils/checkpoint.py (hunk 13 above)
Comment (ronghanghu) [reply pythia/utils/checkpoint.py]: Hi, From Oleksii's experience, without this change, the program frequently crashes (over 50% time) in distributed training due to a race condition, since multiple processes are trying to make directories, and there's no lock on these lines.

File: projects/M4C/scripts/extract_ocr_frcn_feature.py (hunk 7 above)
Comment (ronghanghu) [reply projects/M4C/scripts/extract_ocr_frcn_feature.py]: Yes, there are specific changes for OCR feature extraction. The major change is to allow RoI-Pooling from an externally-specified bounding box (OCR boxes in our use case) instead of from the Faster R-CNN's own RPN proposal. The new branch (https://github.com/ronghanghu/vqa-maskrcnn-benchmark-m4c) is compatible with pythia/scripts/features/extract_features_vmb.py, so can also land it into https://gitlab.com/meetshah1995/vqa-maskrcnn-benchmark.

File: .gitignore (hunk 8 above)
Comment (ronghanghu) [reply .gitignore]: Thanks, I'll remove this change.

The `*-checkpoint.ipynb` files are generated by Jupyter Notebook servers automatically (similar to how vim generates .swp files -- but doesn't delete them after Jupyter Notebook is shut down). But these Jupyter Notebooks were added primarily during our internal analyses and I did not add them here.

**Comment by ronghanghu** [comment]
@apsdehal and @vedanuj Thanks a lot for your review and landing!

I'll take a further look and address the related issues above during this weekend.

**Comment by apsdehal** [comment]
Also, I would suggest you to move to automatic download api instead of making users download everything as per the instructions in your readme.

**Comment by apsdehal** [comment]
Closing as landed internally.
